astropy/units/spectral_quantity.py

```diff
@@ -0,0 +1,291 @@
+        `SpectralQuantity`
+            New spectral coordinate object with data converted to the new unit.
+        """
+
+        # FIXME: need to consider case where doppler equivalency is passed in
+        # equivalencies list, or is u.spectral equivalency is already passed
+
+        if doppler_rest is None:
+            doppler_rest = self._doppler_rest
+
+        if doppler_convention is None:
+            doppler_convention = self._doppler_convention
+        elif doppler_convention not in DOPPLER_CONVENTIONS:
+            raise ValueError("doppler_convention should be one of {0}".format('/'.join(sorted(DOPPLER_CONVENTIONS))))
+
+        if self.unit.is_equivalent(si.km / si.s) and unit.is_equivalent(si.km / si.s):
+
+            # Special case: if the current and final units are both velocity,
+            # and either the rest value or the convention are different, we
+            # need to convert back to frequency temporarily.
+
+            if doppler_convention is not None and self._doppler_convention is None:
+                raise ValueError("Original doppler_convention not set")
+
+            if doppler_rest is not None and self._doppler_rest is None:
+                raise ValueError("Original doppler_rest not set")
+
+            if doppler_rest is None and doppler_convention is None:
+                return super().to(unit, equivalencies=equivalencies)
+            elif (doppler_rest is None) is not (doppler_convention is None):
+                raise ValueError("Either both or neither doppler_rest and "
+                                 "doppler_convention should be defined for "
+                                 "velocity conversions")
+
+            vel_equiv1 = DOPPLER_CONVENTIONS[self._doppler_convention](self._doppler_rest)
+
+            freq = super().to(si.Hz, equivalencies=equivalencies + eq.spectral() + vel_equiv1)
+
+            vel_equiv2 = DOPPLER_CONVENTIONS[doppler_convention](doppler_rest)
+
+            result = freq.to(unit, equivalencies=equivalencies + eq.spectral() + vel_equiv2)
+
+        else:
+
+            additional_equivalencies = eq.spectral()
+
+            if self.unit.is_equivalent(si.km / si.s) or unit.is_equivalent(si.km / si.s):
+
+                if doppler_convention is None:
+                    raise ValueError("doppler_convention not set, cannot convert to/from velocities")
+
+                if doppler_rest is None:
+                    raise ValueError("doppler_rest not set, cannot convert to/from velocities")
+
+                additional_equivalencies += DOPPLER_CONVENTIONS[doppler_convention](doppler_rest)
+
+            result = super().to(unit, equivalencies=equivalencies + additional_equivalencies)
+
+        result._doppler_rest = doppler_rest
+        result._doppler_convention = doppler_convention
+
+        return result
+
+    def to_value(self, *args, **kwargs):
+        return self.to(*args, **kwargs).value
+
+    def _apply_relativistic_doppler_shift(self, velocity):
```

Comment:
I don't think one should define private methods for use of classes in a different scope; I think it may make more sense to move it up to where it is used, or make it a separate function.